## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [6]:
import pandas as pd

client_df = pd.read_csv('Resources/client_dataset.csv')
client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [7]:
# View the column names in the data
client_df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [8]:
# Use the describe function to gather some basic statistics
client_df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [9]:
# Use this space to do any additional research
# and familiarize yourself with the data.

client_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [10]:
client_df['category'].value_counts()
print(client_df['category'].value_counts())

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64


In [11]:
# What three item categories had the most entries?
# Group by 'category' and count entries
category_counts = client_df['category'].value_counts()

# Get the top three categories
top_categories = category_counts.head(3)

# Display the top three item categories
print(top_categories)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64


In [12]:
client_df['subcategory'].value_counts()
print(client_df['subcategory'].value_counts())

subcategory
bathroom supplies     6424
paper products        5874
kitchen supplies      5746
tables                4540
misc                  4286
stands                4201
accounting            3756
project management    3522
plants                3466
wall art              3093
seating               2664
pens                  2603
logistics             1122
telecom                978
audio visual           732
cabling                655
desks                  510
computers              467
Name: count, dtype: int64


In [13]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = category_counts.idxmax()

most_common_category_df = client_df[client_df['category'] == most_common_category]
subcategory_counts = most_common_category_df['subcategory'].value_counts()
most_common_subcategory = subcategory_counts.idxmax()

print(f"The subcategory with the most entries in '{most_common_category}' category is: ")
print(most_common_subcategory, subcategory_counts[most_common_subcategory])


The subcategory with the most entries in 'consumables' category is: 
bathroom supplies 6424


In [14]:
# Which five clients had the most entries in the data?
# Group by 'client_id' and count the number of entries for each client
client_entry_counts = client_df['client_id'].value_counts()

# Get the top 5 clients with the most entries
top_clients = client_entry_counts.head(5)

# Print the results
print("The top 5 clients with the most entries are:")
print(top_clients)


The top 5 clients with the most entries are:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [15]:
# Store the client ids of those top 5 clients in a list.

# Group by 'client_id' and count the number of entries for each client
client_entry_counts = client_df['client_id'].value_counts()

# Get the top 5 clients with the most entries
top_clients = client_entry_counts.head(5)

# Store the client IDs of the top 5 clients in a list
top_client_ids = top_clients.index.tolist()

# Print the list of top client IDs
print("Top 5 client IDs:", top_client_ids)


Top 5 client IDs: [33615, 66037, 46820, 38378, 24741]


In [16]:
# How many total units (the qty column) did the client with the most entries order order?

# Find the client ID with the most entries
client_with_most_entries = client_entry_counts.idxmax()

# Filter the data for this client and sum the 'qty' column
total_units_ordered = client_df[client_df['client_id'] == client_with_most_entries]['qty'].sum()

# Print the result
print(f"Total units ordered by the client with the most entries: {total_units_ordered}")


Total units ordered by the client with the most entries: 64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [17]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
client_df['subtotal'] = client_df['unit_price'] * client_df['qty']

client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [18]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.

# Define a function to calculate shipping cost based on weight
def calculate_shipping(row):
    if row['unit_weight'] > 50:
        return row['unit_weight'] * 7
    else:
        return row['unit_weight'] * 10

# Create a new column 'shipping_cost' that applies the shipping calculation
client_df['shipping_cost'] = client_df.apply(calculate_shipping, axis=1)

# Display the first few rows to verify the new column
client_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_cost  
0           75.0  
1           14.9  
2           16.8  
3           12.3  
4          464.3

In [52]:
client_df['shipping_price_corrected'] = client_df.apply(
    lambda row: row['unit_weight'] * (7 if row['unit_weight'] > 50 else 10) * row['qty'], axis=1
)

In [53]:
client_df['shipping_price_test'] = client_df['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
discrepancies = client_df[client_df['shipping_price'] != client_df['shipping_price_test']]
print(f"Discrepancies in Shipping Price: {discrepancies}")


Discrepancies in Shipping Price: Empty DataFrame
Columns: [first, last, job, phone, email, client_id, order_id, order_date, order_week, order_year, item_id, category, subcategory, unit_price, unit_cost, unit_weight, qty, line_number, subtotal, shipping_cost, total_price, line_cost, line_profit, shipping_price, shipping_price_test, total_price_test, shipping_price_corrected]
Index: []

[0 rows x 27 columns]


In [ ]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
# Create a new column 'total_price' that includes the subtotal, shipping cost, and sales tax
# client_df['total_price'] = client_df['subtotal'] + client_df['shipping_cost']
# client_df['total_price'] *= 1.0925  # Apply 9.25% sales tax

# Display the first few rows to verify the new column
# client_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_cost  \
0     762.71         7.50  105            1  115164.00           75.0   
1      15.09         1.49   21            0     523.95           14.9   
2       7.86         1.68   39            6     527.28           16.8   
3      24.85         1.23   29            3    1056.18           12.3   
4     108.17        46.43   20            1    3902.00          464.3   

     total_price  
0  125898.607500  
1     588.693625  
2     594.407400  
3    1167.314400  
4    4770.182750  

[5 rows x 21 columns]

In [26]:
client_df['total_price'] = (client_df['subtotal'] + client_df['shipping_cost']) * 1.0925
client_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_cost    total_price  line_cost  \
0            1  115164.00           75.0  125898.607500   80159.55   
1            0     523.95           14.9     588.693625     331.79   
2            6     527.28           16.8     594.407400     323.34   
3            3    1056.18           12.3    1167.314400     732.95   
4            1    3902.00          464.3    4770.182750    2627.70   

    line_profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 23 columns]

In [47]:
client_df['total_price_test'] = (client_df['subtotal'] + client_df['shipping_price']) * 1.0925
discrepancies = client_df[client_df['total_price'] != client_df['total_price_test']]
print(f"Discrepancies in Total Price: {discrepancies}")


Discrepancies in Total Price: Empty DataFrame
Columns: [first, last, job, phone, email, client_id, order_id, order_date, order_week, order_year, item_id, category, subcategory, unit_price, unit_cost, unit_weight, qty, line_number, subtotal, shipping_cost, total_price, line_cost, line_profit, shipping_price, shipping_price_test, total_price_test]
Index: []

[0 rows x 26 columns]


In [27]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

# Create a new column 'line_cost' that calculates the cost for each line
client_df['line_cost'] = (client_df['unit_cost'] * client_df['qty']) + client_df['shipping_cost']

# Display the first few rows to verify the new column
client_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_cost    total_price  line_cost  \
0            1  115164.00           75.0  125898.607500   80159.55   
1            0     523.95           14.9     588.693625     331.79   
2            6     527.28           16.8     594.407400     323.34   
3            3    1056.18           12.3    1167.314400     732.95   
4            1    3902.00          464.3    4770.182750    2627.70   

    line_profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 23 columns]

In [28]:
# Create a column for the profit of each line using line cost and line price

# Create a new column 'line_profit' that calculates the profit for each line
client_df['line_profit'] = client_df['total_price'] - client_df['line_cost']

# Display the first few rows to verify the new column
client_df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_cost    total_price  line_cost  \
0            1  115164.00           75.0  125898.607500   80159.55   
1            0     523.95           14.9     588.693625     331.79   
2            6     527.28           16.8     594.407400     323.34   
3            3    1056.18           12.3    1167.314400     732.95   
4            1    3902.00          464.3    4770.182750    2627.70   

    line_profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 23 columns]

In [50]:
for order_id in order_ids:
    print(f"Details for Order ID {order_id}:")
    print(client_df[client_df['order_id'] == order_id][[
        'order_id', 'unit_price', 'qty', 'unit_weight', 'subtotal', 'shipping_price', 'total_price'
    ]])


Details for Order ID 2742071:
       order_id  unit_price  qty  unit_weight   subtotal  shipping_price  \
3446    2742071        5.59   77        10.46     430.43           104.6   
6287    2742071      460.20  260         2.60  119652.00            26.0   
20623   2742071      135.12   12         2.27    1621.44            22.7   
23840   2742071       47.59   22         0.54    1046.98             5.4   
24671   2742071       59.69   24         1.57    1432.56            15.7   
42470   2742071       82.18   44         3.04    3615.92            30.4   

         total_price  
3446      584.520275  
6287   130748.215000  
20623    1796.222950  
23840    1149.725150  
24671    1582.224050  
42470    3983.604600  
Details for Order ID 2173913:
       order_id  unit_price  qty  unit_weight  subtotal  shipping_price  \
627     2173913      110.27   70         2.66   7718.90            26.6   
14089   2173913        3.28   36         1.04    118.08            10.4   
14938   2173913      

In [51]:
for order_id in order_ids:
    total_price_manual = client_df[client_df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id} Total Price (Manual): ${total_price_manual:,.2f}")


Order ID 2742071 Total Price (Manual): $139,844.51
Order ID 2173913 Total Price (Manual): $119,838.36
Order ID 6128929 Total Price (Manual): $752,748.75


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [56]:
# Print the intermediate calculations for the specific order IDs
order_ids = [2742071, 2173913, 6128929]
for order_id in order_ids:
    print(client_df[client_df['order_id'] == order_id][['subtotal', 'shipping_price', 'total_price']])


        subtotal  shipping_price    total_price
3446      430.43           104.6     584.520275
6287   119652.00            26.0  130748.215000
20623    1621.44            22.7    1796.222950
23840    1046.98             5.4    1149.725150
24671    1432.56            15.7    1582.224050
42470    3615.92            30.4    3983.604600
       subtotal  shipping_price   total_price
627     7718.90            26.6   8461.958750
14089    118.08            10.4    140.364400
14938   5835.33            40.8   6419.672025
14962   2145.78            56.7   2406.209400
22004   8655.57             1.7   9458.067475
27713   2660.75            62.7   2975.369125
53155  82306.35            52.2  89976.715875
        subtotal  shipping_price    total_price
3422     8890.75            76.1    9796.283625
13672    5459.60            70.5    6041.634250
14517  117187.29            62.8  128095.723325
16295    3439.92             7.8    3766.634100
21141     731.12            46.0     849.003600
37265   

In [57]:
client_df['subtotal'] = client_df['unit_price'] * client_df['qty']
client_df['shipping_price'] = client_df['unit_weight'].apply(lambda x: x * 7 if x > 50 else x * 10)
client_df['total_price'] = (client_df['subtotal'] + client_df['shipping_price']) * 1.0925


In [58]:
for order_id in order_ids:
    order_df = client_df[client_df['order_id'] == order_id]
    manual_total = ((order_df['subtotal'] + order_df['shipping_price']) * 1.0925).sum()
    print(f"Order ID {order_id} Manual Total: ${manual_total:,.2f}")


Order ID 2742071 Manual Total: $139,844.51
Order ID 2173913 Manual Total: $119,838.36
Order ID 6128929 Manual Total: $752,748.75


In [59]:
print(client_df.duplicated().sum())


0


In [34]:
# Inspect the first few rows with relevant columns
print(client_df[['subtotal','shipping_cost', 'total_price']].head())

# Manually calculate 'total_price' for a specific row
test_row = client_df.iloc[0]  # First row
calculated_price = (test_row['subtotal'] + test_row['shipping_cost']) * 1.0925
print(f"Row 0 calculated total_price: {calculated_price}, Actual: {test_row['total_price']}")


    subtotal  shipping_cost    total_price
0  115164.00           75.0  125898.607500
1     523.95           14.9     588.693625
2     527.28           16.8     594.407400
3    1056.18           12.3    1167.314400
4    3902.00          464.3    4770.182750
Row 0 calculated total_price: 125898.6075, Actual: 125898.6075


In [36]:
# Inspect the relevant columns for anomalies
print(client_df[['unit_price', 'qty', 'subtotal', 'unit_weight', 'shipping_cost']].head())


   unit_price  qty   subtotal  unit_weight  shipping_cost
0     1096.80  105  115164.00         7.50           75.0
1       24.95   21     523.95         1.49           14.9
2       13.52   39     527.28         1.68           16.8
3       36.42   29    1056.18         1.23           12.3
4      195.10   20    3902.00        46.43          464.3


In [37]:
client_df['subtotal'] = client_df['unit_price'] * client_df['qty']
client_df['shipping_price'] = client_df['unit_weight'].apply(
    lambda x: x * 7 if x > 50 else x * 10
)
client_df['total_price'] = (client_df['subtotal'] + client_df['shipping_price']) * 1.0925


In [38]:
print(f"Duplicate rows: {client_df.duplicated().sum()}")
print(f"Missing values:\n{client_df.isnull().sum()}")


Duplicate rows: 0
Missing values:
first             0
last              0
job               0
phone             0
email             0
client_id         0
order_id          0
order_date        0
order_week        0
order_year        0
item_id           0
category          0
subcategory       0
unit_price        0
unit_cost         0
unit_weight       0
qty               0
line_number       0
subtotal          0
shipping_cost     0
total_price       0
line_cost         0
line_profit       0
shipping_price    0
dtype: int64


In [39]:
client_total_prices = client_df.groupby('client_id')['total_price'].sum()
top_5_clients = client_total_prices.nlargest(5)

# Print the results
for client_id, total_price in top_5_clients.items():
    print(f"Client ID: {client_id}, Total Price: ${total_price:,.2f}")


Client ID: 78965, Total Price: $650,096,959.42
Client ID: 85931, Total Price: $226,361,135.01
Client ID: 38473, Total Price: $178,763,956.92
Client ID: 86721, Total Price: $110,986,238.44
Client ID: 84344, Total Price: $101,470,055.85


In [60]:
order_ids = [2742071, 2173913, 6128929]
for order_id in order_ids:
    order_total = client_df[client_df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id} Total Price: ${order_total:,.2f}")


Order ID 2742071 Total Price: $139,844.51
Order ID 2173913 Total Price: $119,838.36
Order ID 6128929 Total Price: $752,748.75


In [32]:
# Verify if 'total_price' is calculated properly
print(client_df[['client_id', 'total_price']].head())

# Re-run grouping and summing
client_total_prices = client_df.groupby('client_id')['total_price'].sum()
top_5_clients = client_total_prices.nlargest(5)

# Print the results
for client_id, total_price in top_5_clients.items():
    print(f"Client ID: {client_id}, Total Price: ${total_price:,.2f}")


   client_id    total_price
0      58515  125898.607500
1      37609     588.693625
2      57113     594.407400
3      46554    1167.314400
4      92089    4770.182750
Client ID: 78965, Total Price: $650,096,959.42
Client ID: 85931, Total Price: $226,361,135.01
Client ID: 38473, Total Price: $178,763,956.92
Client ID: 86721, Total Price: $110,986,238.44
Client ID: 84344, Total Price: $101,470,055.85


In [30]:
# Group by 'client_id' and sum the 'total_price'

client_total_prices = client_df.groupby('client_id')['total_price'].sum()

# Sort by total_price in descending order and get the top 5 clients
top_5_clients = client_total_prices.nlargest(5)

# Print the results
for client_id, total_price in top_5_clients.items():
    print(f"Client ID: {client_id}, Total Price: ${total_price:,.2f}")


Client ID: 78965, Total Price: $650,096,959.42
Client ID: 85931, Total Price: $226,361,135.01
Client ID: 38473, Total Price: $178,763,956.92
Client ID: 86721, Total Price: $110,986,238.44
Client ID: 84344, Total Price: $101,470,055.85


In [31]:
# Filter the dataset for each of the given order IDs
order_2742071 = client_df[client_df['order_id'] == 2742071]
order_2173913 = client_df[client_df['order_id'] == 2173913]
order_6128929 = client_df[client_df['order_id'] == 6128929]

# Sum the 'total_price' column for each order
total_2742071 = order_2742071['total_price'].sum()
total_2173913 = order_2173913['total_price'].sum()
total_6128929 = order_6128929['total_price'].sum()

# Print the results to confirm
print(f"Order ID 2742071 Total Price: ${total_2742071:,.2f}")
print(f"Order ID 2173913 Total Price: ${total_2173913:,.2f}")
print(f"Order ID 6128929 Total Price: ${total_6128929:,.2f}")


Order ID 2742071 Total Price: $139,844.51
Order ID 2173913 Total Price: $119,838.36
Order ID 6128929 Total Price: $752,748.75


In [26]:
# Check your work using the totals above
# List of order IDs and their expected total prices
order_ids_expected_total = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

# Check the total_price for each order ID
for order_id, expected_total in order_ids_expected_total.items():
    actual_total = client_df[client_df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id}: Expected Total = {expected_total}, Actual Total = {actual_total}")


Order ID 2742071: Expected Total = 152811.89, Actual Total = 139844.51202499997
Order ID 2173913: Expected Total = 162388.71, Actual Total = 119838.35704999999
Order ID 6128929: Expected Total = 923441.25, Actual Total = 752748.7454750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


client_id     qty  shipping_price   line_price    line_cost  line_profit
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31

In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      33615   64313             1.828985                  8.377309   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
3      38378   73667             3.429455                 12.906551   
4      24741  239862             5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995